src/transformers/models/whisper/modeling_flax_whisper.py

```diff
@@ -0,0 +1,1365 @@
+            method=_encoder_forward,
+        )
+
+    @add_start_docstrings(WHISPER_DECODE_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=FlaxBaseModelOutputWithPastAndCrossAttentions, config_class=WhisperConfig)
+    def decode(
+        self,
+        decoder_input_ids,
+        encoder_outputs,
+        encoder_attention_mask: Optional[jnp.ndarray] = None,
+        decoder_attention_mask: Optional[jnp.ndarray] = None,
+        decoder_position_ids: Optional[jnp.ndarray] = None,
+        past_key_values: dict = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        train: bool = False,
+        params: dict = None,
+        dropout_rng: PRNGKey = None,
+    ):
+        r"""
+        Returns:
+
+        Example:
+
+        ```python
+        >>> from transformers import WhisperProcessor, FlaxWhisperForConditionalGeneration
+        >>> from datasets import load_dataset
+
+        >>> processor = WhisperProcessor.from_pretrained("openai/whisper-tiny.en")
+        >>> model = FlaxWhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny.en", from_pt=True)
+        >>> ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
+        >>> inputs = processor(ds[0]["audio"]["array"], return_tensors="np")
+        >>> input_features = inputs.input_features
+        >>> encoder_outputs = model.encode(input_features=input_features)
+        >>> decoder_start_token_id = model.config.decoder_start_token_id
+
+        >>> decoder_input_ids = jnp.ones((inputs.input_ids.shape[0], 1), dtype="i4") * decoder_start_token_id
+
+        >>> outputs = model.decode(decoder_input_ids, encoder_outputs)
+        >>> last_decoder_hidden_states = outputs.last_hidden_state
+        ```"""
+
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        encoder_hidden_states = encoder_outputs[0]
+
+        batch_size, sequence_length = decoder_input_ids.shape
+        if decoder_position_ids is None:
+            if past_key_values is not None:
+                raise ValueError("Make sure to provide `decoder_position_ids` when passing `past_key_values`.")
+
+            if decoder_attention_mask is not None:
+                decoder_position_ids = decoder_attention_mask.cumsum(-1) - 1
+            else:
+                decoder_position_ids = jnp.broadcast_to(
+                    jnp.arange(sequence_length)[None, :], (batch_size, sequence_length)
+                )
+
+        if decoder_attention_mask is None:
+            decoder_attention_mask = jnp.ones((batch_size, sequence_length))
```

Comment:
My suggestion would be to copy this as closely as possible from Flax MBart / Bart!
```suggestion
        encoder_hidden_states = encoder_outputs[0]

        batch_size, sequence_length = decoder_input_ids.shape
        if decoder_attention_mask is None:
            decoder_attention_mask = jnp.ones((batch_size, sequence_length))

        if decoder_position_ids is None:
            if past_key_values is not None:
                raise ValueError("Make sure to provide `decoder_position_ids` when passing `past_key_values`.")

            decoder_position_ids = jnp.broadcast_to(
                jnp.arange(sequence_length)[None, :], (batch_size, sequence_length)
            )
```

Reply:
Your suggested change doesn't handle the cases where `attention_mask` is provided and we can use it to compute `decoder_position_ids` that aren't provided - should this case be accounted for here and in other models like Bart and MBart?

Reply:
This isn't really needed for Bart and MBart because they don't use decoder prefixes/expect stuff to be right padded, but I'd argue it's crucial for Whisper to handle multi length decoder prefixes.

Reply:
mBART uses decoder prefix tokens too - we prepend the target label sequence with `[tgt_lang_code]`: https://huggingface.co/docs/transformers/model_doc/mbart#training-of-mbart. The only difference with Whisper is that we can use potentially multiple decoder prefix tokens, rather than a single one in mBART.

And Whisper also uses padding side right? https://github.com/huggingface/transformers/blob/f76518e56a5ef0836a780630de6f5b4456e9aa4a/src/transformers/tokenization_utils_base.py#L1492

=> in terms of the decoder, Whisper operates the same way as BART/mBART (an auto-regressive decoder conditioned on the encoder hidden-states & previously generated tokens). The only change is the multiple prefix tokens, and I don't really see why this necessitates custom behaviour for the `decoder_attention_mask`/`decoder_position_ids`? These should be created in the same way whether the tokens are prefix tokens or non-prefix tokens.

Reply:
Different items in a batch will have different prefix lengths (for example one item in a batch might not be using any previous transcription as context, while another could be using >200 tokens for context). We'll need to add padding for the item that has no previous context to have the same shape as the one with long context.

Reply:
I guess my point is we use `decoder_attention_mask.cumsum(axis=-1) - 1` to compute `decoder_position_ids` in `prepare_inputs_for_generation` for MBART/BART/GPT2/etc so why wouldn't we also use the mask it to compute them in `__call__`?

Reply:
Thanks for the details, I understand your point! Could you please open an issue for this? We can move the discussion over to there (feel free to copy and paste / link your comment from here), as this affects more Flax models than just Whisper!